### PR TITLE
Dicom update API

### PR DESCRIPTION
**Pull request by arunmk-ms**
## Description
New endpoint (/studies/$bulkupdate) for the Dicom update feature.
- Creates an operation id and pass along with the operation href.
- Added feature flag
- Added validation for the input request
- Ensure there are no other operation already running
- Unit and E2E tests added

## Testing
Added unit and E2E tests

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Core/Models/Update/UpdateSpecification.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+namespace Microsoft.Health.Dicom.Core.Models.Update;
+
+public class UpdateSpecification
+{
+    public int PartitionKey { get; set; }
+    public string Id { get; set; }
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Models/Update/UpdateSpecification.cs:11]:
```
public class UpdateSpecification
{
    public IReadOnlyList<string> StudyInstanceUids { get; set; }

    public DicomDataset ChangeDataset { get; set; }
}


```

File: src/Microsoft.Health.Dicom.Core/Features/Store/UpdateInstanceService.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        EnsureArg.IsNotNull(iUrlResolver, nameof(iUrlResolver));
+
+        _guidFactory = guidFactory;
+        _client = client;
+        _urlResolver = iUrlResolver;
+    }
+
+    public async Task<OperationReference> UpdateInstanceAsync(
+        UpdateSpecification spec,
+        CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Store/UpdateInstanceService.cs:52]: We use the partitionKey while calling the durable function. Should we be adding this in the next PR?

File: src/Microsoft.Health.Dicom.Core/Features/Store/UpdateInstanceService.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        EnsureArg.IsNotNull(guidFactory, nameof(guidFactory));
+        EnsureArg.IsNotNull(client, nameof(client));
+        EnsureArg.IsNotNull(iUrlResolver, nameof(iUrlResolver));
+
+        _guidFactory = guidFactory;
+        _client = client;
+        _urlResolver = iUrlResolver;
+    }
+
+    public async Task<OperationReference> UpdateInstanceAsync(
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Store/UpdateInstanceService.cs]: Can we rename this method? It is not exactly updating the instances but queuing an operation to update instance.

File: src/Microsoft.Health.Dicom.Api/Controllers/StoreController.cs [hunk 4]
```diff
@@ -78,6 +87,25 @@ public async Task<IActionResult> PostInstanceInStudyAsync(string studyInstanceUi
         return await PostAsync(studyInstanceUid);
     }
 
+    [HttpPost]
+    [Consumes(KnownContentTypes.ApplicationJson)]
+    [ProducesResponseType((int)HttpStatusCode.OK)]
+    [ProducesResponseType((int)HttpStatusCode.BadRequest)]
+    [ProducesResponseType((int)HttpStatusCode.Conflict)]
```
[the remark this answers is not in the document]
Comment (bcarthic) [reply src/Microsoft.Health.Dicom.Api/Controllers/StoreController.cs:94]: Also since we are going to queue a job, can Not found be returned? We are not going to validate whether the study exists right?

File: src/Microsoft.Health.Dicom.Client/DicomWebClient.Update.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+using EnsureThat;
+using FellowOakDicom;
+using Microsoft.Health.Dicom.Client.Models;
+using Microsoft.Health.Dicom.Core.Models.Update;
+
+namespace Microsoft.Health.Dicom.Client;
+
+public partial class DicomWebClient : IDicomWebClient
+{
+    public async Task<DicomWebResponse> UpdateAsync(
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Client/DicomWebClient.Update.cs:20]: Should we need to be more precise? UpdateStudyAsync?

File: src/Microsoft.Health.Dicom.Core/Features/Audit/AuditEventSubType.cs [hunk 6]
```diff
@@ -55,4 +55,6 @@ public static class AuditEventSubType
     public const string UpdateExtendedQueryTag = "update-extended-query-tag";
 
     public const string BulkImportStore = "bulk-import-store";
+
+    public const string UpdateInstance = "update-instance";
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Audit/AuditEventSubType.cs:59]: I think it is update-study right?

File: src/Microsoft.Health.Dicom.Core/Features/Update/IUpdateInstanceService.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Health.Dicom.Core.Models.Update;
+using Microsoft.Health.Operations;
+
+namespace Microsoft.Health.Dicom.Core.Features.Update;
+
+public interface IUpdateInstanceService
+{
+    public Task<OperationReference> QueueUpdateOperationAsync(
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/IUpdateInstanceService.cs:15]: Can we add some doc on the method?

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceHandler.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        => _updateInstanceService = EnsureArg.IsNotNull(updateInstanceService, nameof(updateInstanceService));
+
+    public async Task<UpdateInstanceResponse> Handle(UpdateInstanceRequest request, CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(request, nameof(request));
+
+        if (await AuthorizationService.CheckAccess(DataActions.Write, cancellationToken) != DataActions.Write)
+            throw new UnauthorizedDicomActionException(DataActions.Write);
+
+        UpdateRequestValidator.ValidateRequest(request);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceHandler.cs:33]: Can we move the validation to the service? I know that we are doing in the handler for others, but it feels right if we do it the service

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        if (request.UpdateSpec == null)
+        {
+            throw new BadRequestException(DicomCoreResource.MissingRequestBody);
+        }
+        else if (request.UpdateSpec.StudyInstanceUids == null)
+        {
+            throw new BadRequestException(DicomCoreResource.MissingRequiredTag);
+        }
+
+        foreach (var StudyInstanceUid in request.UpdateSpec.StudyInstanceUids)
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:37]: What if the StudyInstanceUids is empty?

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        EnsureArg.IsNotNull(dataset, nameof(dataset));
+        foreach (DicomItem item in dataset)
+        {
+            try
+            {
+                item.Validate();
+            }
+            catch (DicomValidationException)
+            {
+                throw new BadRequestException(DicomCoreResource.DicomElementValidationFailed);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:60]: Do you think it will be useful to collect all the errors and then throw?

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs [hunk 11]
```diff
@@ -452,6 +452,17 @@ public async Task GivenInstanceWithPatientIDWithInvalidChars_WhenStoreInstanceWi
         failedAttributesSequence.Items[0].GetString(DicomTag.ErrorComment));
     }
 
+    [Fact]
+    public async Task GivenAValidUpdateAttributeRequest_WhenUpdating_ThenShouldSucceed()
+    {
+        List<string> studyInstanceUIDs = new List<string> { TestUidGenerator.Generate(), TestUidGenerator.Generate() };
+        DicomDataset dataset = new DicomDataset(new DicomPersonName(DicomTag.PatientBirthName, "foo"));
+
+        using DicomWebResponse response = await _instancesManager.UpdateAsync(studyInstanceUIDs, dataset);
```
Comment (bcarthic) [inline test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs:461]: Where do you turn on this feature for E2E tests?

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public async Task<OperationReference> QueueUpdateOperationAsync(
+        UpdateSpecification spec,
+        CancellationToken cancellationToken = default)
+    {
+        EnsureArg.IsNotNull(spec, nameof(spec));
+        EnsureArg.IsNotNull(spec.ChangeDataset, nameof(spec.ChangeDataset));
+
+        UpdateRequestValidator.ValidateDicomDataset(spec.ChangeDataset);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs:63]: Are we missing to validate if the items are in the known tag bucket?

File: src/Microsoft.Health.Dicom.Api/Controllers/StoreController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Api/Controllers/StoreController.cs]: Removed conflict and not found. Added Http Accepted instead of Ok.

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs (hunk 10 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs]: Collected the error for tag and value validation. The errors would be send in the response as below. 
```
Updating the following tags are not supported: '(0010,0010)', '(0010,0011)', '(0010,0012)'
Validation failed for the values of following tags: '(0010,0010)', '(0010,0011)', '(0010,0012)'
```

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs (hunk 11 above)
Comment (arunmk-ms) [reply test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs]: Good catch. I have added it in `InProcTestDicomWebServer`.

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs (hunk 12 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs]: Added the known tag bucket in `UpdateTags.cs`

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs (hunk 9 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs]: There is already a check [here](https://github.com/microsoft/dicom-server/blob/personal/amohankumar/dicomupdateapi/src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs#L32). I have added a more specific error when the `StudyInstanceUids` field is missing or empty

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+        foreach (DicomItem item in dataset)
+        {
+            if (!UpdateTags.UpdateFilterTags.Contains(item.Tag))
+            {
+                invalidTags.Add(item.Tag.ToString());
+            }
+            try
+            {
+                item.Validate();
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:62]: If it is a invalid tag, do we need to validate the value?

Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:62]: Yea, good catch. We dont have to right? I have moved the value validation to `else`

File: src/Microsoft.Health.Dicom.Client/IDicomWebClient.Update.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using FellowOakDicom;
+
+namespace Microsoft.Health.Dicom.Client;
+
+public partial interface IDicomWebClient
+{
+    Task<DicomWebResponse> UpdateStudyAsync(IReadOnlyList<string> studyInstanceUids, DicomDataset dataset, int partitionKey = default, CancellationToken cancellationToken = default);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Client/IDicomWebClient.Update.cs:15]: This needs to be `partitionName`. Users usually dont know what partitionKey is

File: src/Microsoft.Health.Dicom.Client/DicomWebClient.cs [hunk 15]
```diff
@@ -158,6 +158,11 @@ private Uri GenerateWorkitemUpdateRequestUri(string workitemUid, string transact
         return GenerateRequestUri(string.Format(CultureInfo.InvariantCulture, DicomWebConstants.UpdateWorkitemUriFormat, workitemUid, transactionUid), partitionName);
     }
 
+    private Uri GenerateUpdateRequestUri()
+    {
+        return GenerateRequestUri(DicomWebConstants.UpdateAttributeUriString);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Client/DicomWebClient.cs:163]: I don't see partitionName passed as an argument

File: src/Microsoft.Health.Dicom.Core/Features/Audit/AuditEventSubType.cs [hunk 16]
```diff
@@ -55,4 +55,6 @@ public static class AuditEventSubType
     public const string UpdateExtendedQueryTag = "update-extended-query-tag";
 
     public const string BulkImportStore = "bulk-import-store";
+
+    public const string UpdateInstance = "update-study";
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Audit/AuditEventSubType.cs:59]: Variable and the value are not in sync

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+        _guidFactory = guidFactory;
+        _client = client;
+        _urlResolver = iUrlResolver;
+        _contextAccessor = contextAccessor;
+    }
+
+    public async Task<OperationReference> QueueUpdateOperationAsync(
+        UpdateSpecification updateSpecification,
+        CancellationToken cancellationToken = default)
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs:58]: nit: = default not needed

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    public static void ValidateRequest(UpdateSpecification updateSpecification)
+    {
+        EnsureArg.IsNotNull(updateSpecification, nameof(updateSpecification));
+        if (updateSpecification.StudyInstanceUids == null || updateSpecification.StudyInstanceUids.Count == 0)
+        {
+            throw new BadRequestException(string.Format(CultureInfo.CurrentCulture, DicomCoreResource.MissingRequiredField, nameof(updateSpecification.StudyInstanceUids)));
+        }
+        foreach (var StudyInstanceUid in updateSpecification.StudyInstanceUids)
+        {
+            UidValidation.Validate(StudyInstanceUid, nameof(StudyInstanceUid), allowEmpty: true);
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:38]: Why we need to allowEmpty?

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+        if (invalidTags.Count > 0)
+        {
+            errorString = string.Concat(string.Format(CultureInfo.CurrentCulture, DicomCoreResource.DicomUpdateTagValidationFailed, GetCommaSeparatedTags(invalidTags)), "\n");
+        }
+        if (invalidTagValues.Count > 0)
+        {
+            errorString = string.Concat(errorString, string.Format(CultureInfo.CurrentCulture, DicomCoreResource.DicomUpdateTagValueValidationFailed, GetCommaSeparatedTags(invalidTagValues)));
+        }
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:79]: So incase of both invalid tags and invalid values, we are sending either one of them?

File: src/Microsoft.Health.Dicom.Core/Models/Operations/DicomOperation.cs [hunk 20]
```diff
@@ -23,5 +23,10 @@ public enum DicomOperation
     /// <summary>
     /// Specifies an export operation that copies data out of the DICOM server and into an external data store.
     /// </summary>
-    Export
+    Export,
+
+    /// <summary>
+    /// Specifies an update operation that updates the Dicom attributes.
+    /// </summary>
+    Update
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Models/Operations/DicomOperation.cs:31]: I don't see a swagger file generated?

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/DicomInstancesManager.cs [hunk 21]
```diff
@@ -160,4 +160,18 @@ public async Task<DicomWebResponse> QueryWorkitemAsync(string queryString, strin
             .UpdateWorkitemAsync(Enumerable.Repeat(requestDataset, 1), workitemUid, transactionUid, partitionName, cancellationToken)
             .ConfigureAwait(false);
     }
+
+    public async Task<DicomWebResponse> UpdateStudyAsync(
+        IReadOnlyList<string> studyInstanceUids,
+        DicomDataset dataset,
+        int partitionKey = default,
```
Comment (bcarthic) [inline test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/DicomInstancesManager.cs:167]: partitionName?

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs [hunk 22]
```diff
@@ -452,6 +452,17 @@ public async Task GivenInstanceWithPatientIDWithInvalidChars_WhenStoreInstanceWi
         failedAttributesSequence.Items[0].GetString(DicomTag.ErrorComment));
     }
 
+    [Fact]
+    public async Task GivenAValidUpdateAttributeRequest_WhenUpdating_ThenShouldSucceed()
```
Comment (bcarthic) [inline test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs:456]: Could we add the same test in `DataPartitionEnabledTests`?

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs (hunk 19 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:85]: No, we will send both the tag validation and tag value validation errors together.

Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:85]: Sample response would be as below.
```
Updating the following tags are not supported: '(0010,1111)'
Validation failed for the values of following tags: '(0010,0010)'
```

File: src/Microsoft.Health.Dicom.Client/IDicomWebClient.Update.cs (hunk 14 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Client/IDicomWebClient.Update.cs]: Updated to `partitionName`

File: src/Microsoft.Health.Dicom.Client/DicomWebClient.cs (hunk 15 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Client/DicomWebClient.cs]: Added `partitionName` as argument

File: src/Microsoft.Health.Dicom.Core/Features/Audit/AuditEventSubType.cs (hunk 16 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Audit/AuditEventSubType.cs]: Corrected the same

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+
+        OperationReference activeOperation = await _client
+            .FindOperationsAsync(Query, cancellationToken)
+            .FirstOrDefaultAsync(cancellationToken);
+
+        if (activeOperation != null)
+            throw new ExistingUpdateOperationException(activeOperation);
+
+        var operationId = _guidFactory.Create();
+        var partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
```
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs]: Removed

File: src/Microsoft.Health.Dicom.Core/Models/Operations/DicomOperation.cs (hunk 20 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Models/Operations/DicomOperation.cs:31]: Please check now. I have updated the new swagger

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs (hunk 22 above)
Comment (arunmk-ms) [reply test/Microsoft.Health.Dicom.Web.Tests.E2E/Rest/StoreTransactionTests.cs:456]: Done

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs (hunk 17 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs]: Removed

File: src/Microsoft.Health.Dicom.Client/Models/UpdateSpecification.cs [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using FellowOakDicom;
+
+namespace Microsoft.Health.Dicom.Core.Models.Update;
+
+public class UpdateSpecification
+{
+    public UpdateSpecification(IReadOnlyList<string> studyInstanceUids, DicomDataset changeDataset)
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Client/Models/UpdateSpecification.cs:13]: nit: Do we need this constructor, if so can we make the property to only get?

File: src/Microsoft.Health.Dicom.Core/Features/Update/IUpdateInstanceService.cs [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+using Microsoft.Health.Dicom.Core.Models.Update;
+using Microsoft.Health.Operations;
+
+namespace Microsoft.Health.Dicom.Core.Features.Update;
+
+/// <summary>
+/// Provides functionality to queue the operation for updating the study attributes
+/// for a list of studyInstanceUids.
+/// </summary>
+public interface IUpdateInstanceService
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/IUpdateInstanceService.cs:17]: nit: Could we update to IUpdateOperationInstanceService

Since we cannot have the queuing and the update in the same service

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+using Microsoft.Health.Dicom.Core.Models.Update;
+
+namespace Microsoft.Health.Dicom.Core.Features.Update;
+
+/// <summary>
+/// Provides functionality to validate an <see cref="UpdateInstanceRequest"/>.
+/// </summary>
+public static class UpdateRequestValidator
+{
+    private const int MaxStudyInstanceUids = 50;
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateRequestValidator.cs:23]: nit: We can move this to UpdateTags. UpdateTags class could be like `QueryLimit`, where it contains the filters and constraints

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateTags.cs [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using FellowOakDicom;
+
+namespace Microsoft.Health.Dicom.Core.Features.Update;
+
+internal static class UpdateTags
+{
+    public static readonly HashSet<DicomTag> UpdateFilterTags = new HashSet<DicomTag>()
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateTags.cs:13]: Since this is Study Update tags, we can name as UpdateStudyFilterTags

File: src/Microsoft.Health.Dicom.Core/Models/Update/UpdateSpecification.cs [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using FellowOakDicom;
+
+namespace Microsoft.Health.Dicom.Core.Models.Update;
+
+public class UpdateSpecification
+{
+    public UpdateSpecification(IReadOnlyList<string> studyInstanceUids, DicomDataset changeDataset)
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Models/Update/UpdateSpecification.cs:13]: Same comment as before

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/DicomInstancesManager.cs [hunk 29]
```diff
@@ -1,4 +1,4 @@
-﻿// -------------------------------------------------------------------------------------------------
```
Comment (bcarthic) [inline test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/DicomInstancesManager.cs:1]: Revert this change?

File: src/Microsoft.Health.Dicom.Core/Features/Validation/UidValidation.cs [hunk 30]
```diff
@@ -1,4 +1,4 @@
-﻿// -------------------------------------------------------------------------------------------------
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Validation/UidValidation.cs:1]: Revert?

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/Update/UpdateRequestValidatorTests.cs [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+                { DicomTag.PatientID, "123" },
+                { DicomTag.PatientName, "Anonymous" }
+            } };
+    }
+
+    public static IEnumerable<object[]> GetInvalidDicomDataset()
+    {
+        yield return new object[] { new DicomDataset(new DicomPersonName(DicomTag.IssuerOfPatientID, "Issuer")) };
+    }
+}
```
[the remark this answers is not in the document]
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core.UnitTests/Features/Update/UpdateRequestValidatorTests.cs:101]: Yes, line number 78 where I added around 60 studyInstanceUids to reach the max count

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/DicomInstancesManager.cs (hunk 29 above)
Comment (arunmk-ms) [reply test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/DicomInstancesManager.cs:1]: git is not allowing to rollback this to previous commit.

File: src/Microsoft.Health.Dicom.Core/Features/Validation/UidValidation.cs (hunk 30 above)
Comment (arunmk-ms) [reply src/Microsoft.Health.Dicom.Core/Features/Validation/UidValidation.cs:1]: git is not allowing to rollback this to previous commit.

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateTags.cs [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using FellowOakDicom;
+
+namespace Microsoft.Health.Dicom.Core.Features.Update;
+
+internal static class UpdateTags
+{
+    public const int QueryLimit = 50;
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateTags.cs:13]: I believe this is MaxStudyInstanceUidLimit

File: src/Microsoft.Health.Dicom.Core/Features/Update/UpdateTags.cs [hunk 33]
```diff
@@ -0,0 +1,50 @@
...
+
+    public static readonly HashSet<DicomTag> UpdateStudyFilterTags = new HashSet<DicomTag>()
```
Comment (bcarthic) [inline src/Microsoft.Health.Dicom.Core/Features/Update/UpdateTags.cs:15]: nit: internal

**Review by bcarthic — APPROVED** [review]
